pkg/packetfilter/nftables/rule_conversion.go

```diff
@@ -0,0 +1,229 @@
+	}
+
+	return ruleSpec
+}
+
+func mssClampToRuleSpec(ruleSpec []string, clampType packetfilter.MssClampType, mssValue string) []string {
+	switch clampType {
+	case packetfilter.UndefinedMSS:
+	case packetfilter.ToPMTU:
+		ruleSpec = append(ruleSpec, "size", "set", "rt", "mtu")
+	case packetfilter.ToValue:
+		ruleSpec = append(ruleSpec, "size", "set", mssValue)
+	}
+
+	return ruleSpec
+}
+
+func setToRuleSpec(ruleSpec []string, srcSetName, destSetName string) []string {
+	if srcSetName != "" {
+		ruleSpec = append(ruleSpec, "ip", "saddr", "@"+srcSetName)
+	}
+
+	if destSetName != "" {
+		ruleSpec = append(ruleSpec, "ip", "daddr", "@"+destSetName)
+	}
+
+	return ruleSpec
+}
+
+func toNftRuleSpec(rule *packetfilter.Rule) string {
+	ruleSpec := protoToRuleSpec([]string{}, rule.Proto, rule.DPort)
+
+	if rule.SrcCIDR != "" {
+		ruleSpec = append(ruleSpec, "ip", "saddr", rule.SrcCIDR)
+	}
+
+	if rule.DestCIDR != "" {
+		ruleSpec = append(ruleSpec, "ip", "daddr", rule.DestCIDR)
+	}
+
+	if rule.MarkValue != "" && rule.Action != packetfilter.RuleActionMark {
+		// syntax for MarkValue '0xc0000': meta mark & 0xc0000 == 0xc0000
+		ruleSpec = append(ruleSpec, "meta", "mark", "&", rule.MarkValue, "==", rule.MarkValue)
+	}
+
+	ruleSpec = setToRuleSpec(ruleSpec, rule.SrcSetName, rule.DestSetName)
+
+	if rule.OutInterface != "" {
+		ruleSpec = append(ruleSpec, "oifname", rule.OutInterface)
+	}
+
+	if rule.InInterface != "" {
+		ruleSpec = append(ruleSpec, "iifname", rule.InInterface)
+	}
+
+	if rule.Action == packetfilter.RuleActionMss {
+		ruleSpec = append(ruleSpec, "tcp", "flags", "syn / syn,rst")
+	}
+
+	ruleSpec = append(ruleSpec, "counter")
+	ruleSpec = append(ruleSpec, ruleActionToStr[rule.Action]...)
+
+	if rule.Action == packetfilter.RuleActionJump {
+		ruleSpec = append(ruleSpec, rule.TargetChain)
+	}
+
+	if rule.SnatCIDR != "" {
+		ruleSpec = append(ruleSpec, "to", rule.SnatCIDR)
+	}
+
+	if rule.DnatCIDR != "" {
+		ruleSpec = append(ruleSpec, "to", rule.DnatCIDR)
+	}
+
+	ruleSpec = mssClampToRuleSpec(ruleSpec, rule.ClampType, rule.MssValue)
+
+	if rule.MarkValue != "" && rule.Action == packetfilter.RuleActionMark {
+		// syntax for MarkValue '0xc0000': set mark | 0xc0000
+		ruleSpec = append(ruleSpec, "set", "mark", "|", rule.MarkValue)
+	}
+
+	str := strings.Join(ruleSpec, " ")
+
+	logger.V(log.TRACE).Infof("toNftRuleSpec: from %q to %q", rule, str)
+
+	return str
+}
+
+func mustWriteString(buf *bytes.Buffer, s string) {
+	_, err := buf.WriteString(s + " ")
+	utilruntime.Must(err)
+}
+
+func mustWriteUint32(buf *bytes.Buffer, i uint32) {
+	mustWriteString(buf, strconv.FormatInt(int64(i), 10))
+}
+
+func SerializeRule(rule *packetfilter.Rule) string {
+	var buf bytes.Buffer
+
+	_, err := buf.WriteString(serializedVersion)
+	utilruntime.Must(err)
+
+	mustWriteUint32(&buf, uint32(rule.Action))
+	mustWriteUint32(&buf, uint32(rule.ClampType))
+	mustWriteUint32(&buf, uint32(rule.Proto))
+
+	mustWriteString(&buf, rule.SrcSetName)
```

Comment:
Why do we want to panic here ?  don't we want to return an error and let the caller decide how to handle this err ?

Reply:
`buf.WriteString` documents that the return "_err is always nil. If the buffer becomes too large, WriteString will panic with [ErrTooLarge]._". So I didn't want to pollute the code with a lot of unnecessary error checking. The other option is to ignore the return error instead of panicking. I'm fine either way.

We could get an error in `Deserialize` while reading strings and parsing to integers. I did do error checking and error propagation when parsing the serialized version. Not sure how this could happen outside of someone manually entering a rule out-of-band in the submariner table. We could just log the error and not propagate. After successfully reading the serialized version, if a parsing error occurs after that, I panic b/c that means there's serious corruption in nft.

Reply:
if buf.WriteString always returns nil or panic with ErrTooLarge , maybe we should ignore the return error as you suggeseted.